File: ruby/lib/keycloak_sdk/auth_client.rb
# frozen_string_literal: true

require "rack/oauth2"
require "securerandom"
require "digest"
require "base64"

module KeycloakSdk
  # 인증 파사드. rack-oauth2를 래핑(그랜트·PKCE)하고 introspection(RFC7662)·logout은 Faraday로 손수 수행한다.
  # TokenProvider를 구현하지만(직접 사용용), admin은 캐싱 ClientCredentialsTokenProvider를 별도로 쓴다(§4).
  class AuthClient
    include TokenProvider

    def initialize(config:, http:, jwt_validator:)
      @config = config
      @http = http
      @jwt_validator = jwt_validator
      @endpoints = OidcEndpoints.from_config(config)
      configure_rack_oauth2_timeouts(config)
    end

    def create_authorization_request(redirect_uri:, scopes: nil, state: SecureRandom.urlsafe_base64(24),
                                     nonce: SecureRandom.urlsafe_base64(24))
      verifier = SecureRandom.urlsafe_base64(64)
      challenge = Base64.urlsafe_encode64(Digest::SHA256.digest(verifier), padding: false)
      params = {
        scope: (scopes || @config.scopes).join(" "),
        state: state,
        nonce: nonce,
        code_challenge: challenge,
        code_challenge_method: :S256
      }
      url = oauth_client(redirect_uri: redirect_uri).authorization_uri(params)
      AuthorizationRequest.new(url: url.to_s, state: state, code_verifier: verifier, nonce: nonce)
    end

    # `expected_nonce`가 주어지면(create_authorization_request가 항상 돌려주는 nonce) 응답 id_token을
    # realm JWKS로 서명·iss·aud·exp까지 강화 검증한 뒤 nonce 클레임을 대조한다 — OIDC nonce 재생
    # 방지. 불일치·부재·검증실패는 모두 거부(fail-closed). 생략 시 id_token 검증을 건너뛴다
    # (여덟 언어 공통 — exchange에서 nonce를 필수로 만들지 않는다).
    def exchange_code(code:, code_verifier:, redirect_uri:, expected_nonce: nil)
      client = oauth_client(redirect_uri: redirect_uri)
      client.authorization_code = code
      token_set = token_request("authorization_code exchange") { client.access_token!(code_verifier: code_verifier) }
      verify_nonce!(token_set.id_token, expected_nonce) unless expected_nonce.nil?
      token_set
    end

    def refresh(refresh_token:)
      client = oauth_client
      client.refresh_token = refresh_token
      token_request("refresh") { client.access_token! }
    end

    def client_credentials_token
      token_request("client-credentials") { oauth_client.access_token!(scope: @config.scopes.join(" ")) }
    end

    # TokenProvider 계약(직접 사용용). admin은 캐싱 provider를 별도로 쓴다.
    def access_token
      client_credentials_token.access_token
    end

    def introspect(token)
      resp = @http.post(@endpoints.introspection, {
                          token: token, client_id: @config.client_id, client_secret: @config.client_secret
                        })
      raise AuthError, "introspection failed: HTTP #{resp.status}" unless resp.success?

      IntrospectionResult.from_response(resp.body)
    rescue Faraday::Error => e
      raise TransportError, "introspection transport error: #{RedactedCause.describe(e)}", cause: RedactedCause.new(e)
    end

    def logout(refresh_token:)
      resp = @http.post(@endpoints.end_session, {
                          client_id: @config.client_id, client_secret: @config.client_secret,
                          refresh_token: refresh_token
                        })
      raise AuthError, "logout failed: HTTP #{resp.status}" unless resp.success?

      nil
    rescue Faraday::Error => e
      raise TransportError, "logout transport error: #{RedactedCause.describe(e)}", cause: RedactedCause.new(e)
    end

    def validate(token)
      @jwt_validator.validate(token)
    end

    private

    # rack-oauth2의 프로세스 전역 HTTP 타임아웃을 Config로 설정한다(require 시점 하드코딩 대신).
    # 타임아웃은 Faraday::Connection이 아니라 그 #options(Faraday::RequestOptions)에 있다
    # (Connection에 open_timeout=/timeout= 세터가 없어 NoMethodError — 게차 참조).
    def configure_rack_oauth2_timeouts(config)
      Rack::OAuth2.http_config do |conn|
        conn.options.open_timeout = config.connect_timeout
        conn.options.timeout = config.read_timeout
      end
    end

    # id_token의 nonce 클레임을 대조하기 전에 강화 JwtValidator로 서명·iss·aud·exp까지 검증한다
    # (액세스 토큰과 id_token 모두 aud=client_id이므로 검증기를 공유해도 안전 — Kotlin/.NET 동형).
    # ⚠️ `config.expected_audience`를 설정하면 이 공유 검증기가 id_token에도 그 값을 요구한다 —
    # 이 흐름을 쓴다면 해당 오디언스를 id_token에도 매핑해야 한다(audience 매퍼의 "Add to ID token").
    def verify_nonce!(id_token, expected_nonce)
      raise AuthError, "authorization_code exchange failed: missing id_token for nonce validation" if id_token.nil?

      validated = @jwt_validator.validate(id_token)
      return if validated.claims["nonce"] == expected_nonce

      raise AuthError, "authorization_code exchange failed: unexpected nonce"
    rescue TokenValidationError => e
      raise AuthError, "authorization_code exchange failed: invalid id_token: #{e.message}"
    end

    # rack-oauth2 토큰 호출의 오류 경계(§4). 하위 예외는 SDK 타입이 되고 `cause` 에는 원본 대신 `RedactedCause` 가 달린다.
    # ⚠️ OAuth 오류는 **코드와 HTTP 상태만** 싣는다 — `error_description` 은 서버가 고른 자유 문장이라 토큰을
    # 되울릴 수 있고, 오류 본문이 JSON 이 아니면 rack-oauth2 가 본문 전체를 거기 넣는다.
    # ⚠️ 마지막 `StandardError` 는 rack-oauth2 가 형식이 틀린 200 을 읽다 내는 것(NoMethodError·AttrMissing·
    # 'Unknown Token Type')과 `to_token_set` 의 형 변환 실패다 — 원본이 새면 Ruby 3.2 의 NoMethodError 가 본문을 인용한다.
    def token_request(operation)
      to_token_set(yield)
    rescue Error
      raise
    rescue Rack::OAuth2::Client::Error => e
      code = e.response[:error].is_a?(String) ? e.response[:error] : ""
      raise AuthError.new([operation, "failed:", code, "(HTTP #{e.status})"].reject(&:empty?).join(" "),
                          oauth_error: code), cause: RedactedCause.new(e)
    rescue Faraday::Error => e
      raise TransportError, "token endpoint transport error: #{RedactedCause.describe(e)}", cause: RedactedCause.new(e)
    rescue StandardError => e
      raise AuthError, "#{operation} failed: unusable token response (#{e.class})", cause: RedactedCause.new(e)
    end

    def oauth_client(redirect_uri: nil)
      Rack::OAuth2::Client.new(
        identifier: @config.client_id,
        secret: @config.client_secret,
        authorization_endpoint: @endpoints.authorization,
        token_endpoint: @endpoints.token,
        redirect_uri: redirect_uri
      )
    end

    # `expires_in` 은 `TokenSet.from_response` 와 같이 정수로 읽는다(문자열 "300" 허용) — 예전에는 문자열이면
    # `Float + String` 의 TypeError 가 경계를 뚫었다. 변환 실패는 `token_request` 가 AuthError 로 바꾼다.
    def to_token_set(token)
      raw = token.raw_attributes || {}
      expires_in = token.expires_in && Integer(token.expires_in)
      TokenSet.new(
        access_token: token.access_token,
        token_type: "Bearer",
        expires_in: expires_in,
        refresh_token: token.refresh_token,
        id_token: raw[:id_token] || raw["id_token"],
        scope: raw[:scope] || raw["scope"],
        expires_at: expires_in ? Time.now.to_f + expires_in : nil
      )
    end
  end
end

File: php/src/AuthClient.php
<?php

declare(strict_types=1);

namespace Xzawed\Keycloak;

use GuzzleHttp\ClientInterface as GuzzleClientInterface;
use GuzzleHttp\Exception\ConnectException;
use GuzzleHttp\Exception\GuzzleException;
use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
use League\OAuth2\Client\Token\AccessToken;
use Xzawed\Keycloak\Exception\KeycloakAuthError;
use Xzawed\Keycloak\Exception\KeycloakException;
use Xzawed\Keycloak\Exception\KeycloakTransportError;
use Xzawed\Keycloak\Exception\SanitizedCause;
use Xzawed\Keycloak\Exception\TokenValidationError;
use Xzawed\Keycloak\Internal\OAuthErrorCode;
use Xzawed\Keycloak\Internal\PkceKeycloakProvider;
use Xzawed\Keycloak\Token\AuthorizationRequest;
use Xzawed\Keycloak\Token\IntrospectionResult;
use Xzawed\Keycloak\Token\TokenSet;
use Xzawed\Keycloak\Token\ValidatedToken;

/**
 * 인증 파사드 — league/oauth2-client + stevenmaguire/oauth2-keycloak을 감싸고
 * introspect/logout(RFC 7662 + 백채널)은 손수 구현한다. validate()는 T7 JwtValidator에 위임.
 *
 * 네트워크 경계 모듈 — 커버리지 게이트 omit(phpunit.xml). 전체 흐름은 Task 11 통합테스트로 검증.
 */
final class AuthClient
{
    private PkceKeycloakProvider $provider;

    public function __construct(
        private readonly KeycloakConfig $config,
        private readonly OidcEndpoints $endpoints,
        private readonly JwtValidator $validator,
        private readonly GuzzleClientInterface $http,
    ) {
        $this->provider = new PkceKeycloakProvider([
            'authServerUrl' => $config->serverUrl,
            'realm' => $config->realm,
            'clientId' => $config->clientId,
            'clientSecret' => $config->clientSecret ?? '',
            'redirectUri' => $config->redirectUri ?? '',
            'version' => '26.0.0',   // >=20 → openid 스코프, >=18 → logout id_token_hint
        ], ['httpClient' => $http]);
    }

    /**
     * @param string|null $redirectUri 이 요청에만 쓸 콜백 URL. null이면 config 값.
     *   ⚠️ 콜백이 여럿인 앱(멀티테넌트·환경별)이 클라이언트 하나로 그것을 섬기기 위한 자리다 —
     *   나머지 여덟 SDK와 동형. **`exchangeCode()`에 같은 값을 넘겨야 한다**(RFC 6749 §4.1.3).
     */
    public function createAuthorizationRequest(?string $redirectUri = null): AuthorizationRequest
    {
        // league/oauth2-client getAuthorizationUrl(['nonce' => $n])는 쿼리에 nonce=를 그대로 싣는다
        // (실측 2026-08-15: HAS_NONCE_KEY=yes MATCHES=yes). pkceMethod 생성자 옵션의 no-op과 다른 부류.
        $nonce = self::randomUrlSafe();
        $options = [
            'scope' => implode(' ', $this->config->scopes),
            'nonce' => $nonce,
        ];
        // ⚠️ 키가 비어 있으면 상류가 생성자 값으로 채운다(`AbstractProvider::getAuthorizationParameters`).
        // 그래서 null일 때는 **키 자체를 넣지 않는다** — 빈 문자열을 넣으면 config 폴백이 죽는다.
        if ($redirectUri !== null) {
            $options['redirect_uri'] = $redirectUri;
        }
        $url = $this->provider->getAuthorizationUrl($options);
        $verifier = $this->provider->getPkceCode();

        return new AuthorizationRequest(
            url: $url,
            state: $this->provider->getState(),
            codeVerifier: (string) $verifier,
            nonce: $nonce,
        );
    }

    /**
     * Authorization-code + PKCE 토큰 교환. 콜백에서 OAuth `state`를 createAuthorizationRequest()가 발급한
     * 값과 대조하는 것은 호출자 책임이다(SDK는 무상태라 state를 검증하지 않는다 — Node/Go/C# SDK와 동형).
     *
     * `$expectedNonce`가 주어지면(createAuthorizationRequest()가 돌려준 nonce) 응답 id_token을
     * JwtValidator로 서명·iss·aud·exp까지 검증한 뒤 nonce 클레임을 대조한다 — OIDC nonce 재생
     * 방지. 불일치·부재·검증실패는 모두 거부(fail-closed). null이면 id_token 검증을 건너뛴다
     * (여덟 언어 공통 패턴 — 필수로 만들지 않는다).
     */
    public function exchangeCode(
        #[\SensitiveParameter] string $code,
        #[\SensitiveParameter] string $codeVerifier,
        ?string $expectedNonce = null,
        ?string $redirectUri = null,
    ): TokenSet {
        $this->provider->setPkceCode($codeVerifier);
        $options = ['code' => $code];
        // ⚠️ 인가 때 쓴 값과 **같아야** 한다(RFC 6749 §4.1.3) — 다르면 Keycloak이 거부한다.
        // null일 때 키를 넣지 않는 이유는 createAuthorizationRequest()와 같다(상류 폴백 보존).
        if ($redirectUri !== null) {
            $options['redirect_uri'] = $redirectUri;
        }
        $tokens = $this->toTokenSet($this->getAccessToken('authorization_code', $options));
        if ($expectedNonce !== null) {
            $this->requireValidNonce($tokens->idToken, $expectedNonce);
        }

        return $tokens;
    }

    /**
     * id_token의 nonce 클레임을 대조하기 전에 강화 JwtValidator로 서명·iss·aud·exp까지 검증한다.
     * 거부는 자매 언어와 같이 Auth 계급(KeycloakAuthError)이다 — TokenValidationError는
     * validator가 던지고 여기서 감싼다(Ruby AuthError 동형).
     */
    private function requireValidNonce(#[\SensitiveParameter] ?string $idToken, string $expectedNonce): void
    {
        if ($idToken === null || $idToken === '') {
            throw new KeycloakAuthError('authorization code exchange failed: missing id_token for nonce validation');
        }
        try {
            $validated = $this->validator->validate($idToken);
        } catch (TokenValidationError $e) {
            throw new KeycloakAuthError('authorization code exchange failed: invalid id_token: ' . $e->getMessage(), previous: $e);
        }
        $actual = $validated->claims['nonce'] ?? null;
        if (!is_string($actual) || $actual !== $expectedNonce) {
            throw new KeycloakAuthError('authorization code exchange failed: unexpected nonce');
        }
    }

    /** state/PKCE와 같은 강도의 URL-safe 난수(base64url, 패딩 없음). */
    private static function randomUrlSafe(): string
    {
        return rtrim(strtr(base64_encode(random_bytes(24)), '+/', '-_'), '=');
    }

    public function clientCredentialsToken(): TokenSet
    {
        // config.scopes를 client-credentials 요청에 전달한다(authz-url과 동형). 누락 시 커스텀 스코프가
        // 무시돼 언더스코프 토큰이 발급된다(다른 SDK는 모두 threading — Node/Rust token-provider 동형).
        $options = $this->config->scopes === [] ? [] : ['scope' => implode(' ', $this->config->scopes)];

        return $this->toTokenSet($this->getAccessToken('client_credentials', $options));
    }

    public function refresh(#[\SensitiveParameter] string $refreshToken): TokenSet
    {
        return $this->toTokenSet($this->getAccessToken('refresh_token', ['refresh_token' => $refreshToken]));
    }

    public function validate(#[\SensitiveParameter] string $accessToken): ValidatedToken
    {
        return $this->validator->validate($accessToken);
    }

    public function introspect(#[\SensitiveParameter] string $token): IntrospectionResult
    {
        // RFC 7662 — league/stevenmaguire 미제공, 손수 POST(client_secret_basic)
        // RFC 6749 §2.3.1: Basic 자격증명은 각 구성요소를 먼저 percent-encode한다
        // (secret에 ':'/예약문자가 있어도 자격증명이 깨지지 않도록).
        $basic = base64_encode(rawurlencode($this->config->clientId) . ':' . rawurlencode($this->config->clientSecret ?? ''));
        try {
            $response = $this->http->request('POST', $this->endpoints->introspection(), [
                'headers' => ['Authorization' => 'Basic ' . $basic, 'Content-Type' => 'application/x-www-form-urlencoded'],
                'form_params' => ['token' => $token, 'token_type_hint' => 'access_token'],
            ]);
        } catch (ConnectException $e) {
            throw new KeycloakTransportError('introspection unreachable', previous: SanitizedCause::of($e));
        } catch (GuzzleException $e) {
            throw new KeycloakAuthError('introspection failed', previous: SanitizedCause::of($e));
        } catch (KeycloakException $e) {
            throw $e;
        } catch (\Throwable $e) {
            throw new KeycloakTransportError('introspection failed unexpectedly', previous: SanitizedCause::of($e));
        }
        $json = json_decode((string) $response->getBody(), true);
        if (!is_array($json)) {
            throw new KeycloakAuthError('introspection returned non-JSON');
        }

        return IntrospectionResult::fromArray(self::stringKeyed($json));
    }

    public function logoutUrl(TokenSet $tokens): string
    {
        $token = new AccessToken(['access_token' => $tokens->accessToken, 'id_token' => $tokens->idToken]);

        return $this->provider->getLogoutUrl(['access_token' => $token]);
    }

    public function logout(#[\SensitiveParameter] string $refreshToken): void
    {
        // 백채널 end_session POST(refresh_token + client creds) — league/stevenmaguire 미제공
        try {
            $this->http->request('POST', $this->endpoints->endSession(), [
                'form_params' => [
                    'client_id' => $this->config->clientId,
                    'client_secret' => $this->config->clientSecret ?? '',
                    'refresh_token' => $refreshToken,
                ],
            ]);
        } catch (ConnectException $e) {
            throw new KeycloakTransportError('logout unreachable', previous: SanitizedCause::of($e));
        } catch (GuzzleException $e) {
            throw new KeycloakAuthError('logout failed', previous: SanitizedCause::of($e));
        } catch (KeycloakException $e) {
            throw $e;
        } catch (\Throwable $e) {
            throw new KeycloakTransportError('logout failed unexpectedly', previous: SanitizedCause::of($e));
        }
    }

    /**
     * ⚠️ `$options` 는 refresh_token·authorization code 를 쥔다 — 이 프레임이 SDK 예외 트레이스의 #0 이다.
     * ⚠️ 하위 예외는 원본이 아니라 `SanitizedCause` 사본으로 단다(그 클래스의 docblock).
     *
     * @param array<string,mixed> $options
     */
    private function getAccessToken(string $grant, #[\SensitiveParameter] array $options = []): AccessToken
    {
        try {
            $token = $this->provider->getAccessToken($grant, $options);
        } catch (IdentityProviderException $e) {
            $body = $e->getResponseBody();
            // ⚠️ league 의 메시지는 `error: error_description` 이다 — 설명은 응답 본문이라 IdP 가 토큰을 되울리면
            // 그대로 찍혔다(실측). 메시지에는 OAuth `error` 코드만 싣고(`oauthError` 와 같은 값), 그 코드도 코드
            // 모양일 때만 받는다 — `error` 자리에 토큰을 실어 보내면 공개 프로퍼티째 찍혔다(Grok 레그 실측).
            $oauth = OAuthErrorCode::of(is_array($body) ? ($body['error'] ?? null) : null);

            throw new KeycloakAuthError('token request rejected' . ($oauth === null ? '' : ': ' . $oauth), oauthError: $oauth, previous: SanitizedCause::of($e));
        } catch (ConnectException $e) {
            throw new KeycloakTransportError('token endpoint unreachable', previous: SanitizedCause::of($e));
        } catch (GuzzleException $e) {
            throw new KeycloakTransportError('token request failed', previous: SanitizedCause::of($e));
        } catch (\UnexpectedValueException $e) {
            // league는 토큰 엔드포인트가 비-JSON/파싱불가 응답을 줄 때 SPL \UnexpectedValueException을
            // 던진다(IdentityProviderException이 아님 — OAuth 에러 바디가 아니라 응답 자체가 깨진 경우).
            // JwksStore가 동일 상황(비-JSON 응답)을 KeycloakTransportError로 매핑하는 것과 동형.
            throw new KeycloakTransportError('token endpoint returned unexpected response', previous: SanitizedCause::of($e));
        } catch (\Throwable $e) {
            // league/oauth2-client 및 그 하위 의존성이 던질 수 있는 그 밖의 미분류 예외까지 전부
            // 여기로 수렴시켜 "getAccessToken을 벗어나는 미분류 하위 예외는 없다" 경계를 보장한다.
            // (KeycloakException pass-through 분기는 넣지 않는다 — $this->provider->getAccessToken()은
            // league의 완전히 타입드된 메서드라 PHPStan이 우리 자신의 예외가 여기서 절대 던져지지 않음을
            // 증명하므로 그 분기는 도달불가 dead catch로 확정 보고된다: catch.neverThrown.)
            throw new KeycloakTransportError('token request failed', previous: SanitizedCause::of($e));
        }
        if (!$token instanceof AccessToken) {
            throw new KeycloakAuthError('unexpected access token implementation');
        }

        return $token;
    }

    /**
     * ⚠️ league 는 `access_token`·`refresh_token` 의 **존재**만 보고(`empty`) 타입은 안 본다 — 숫자·객체가 그대로
     * 실려 `strict_types` 의 `\TypeError` 가 공개 API 로 샜고, 그 트레이스 인자(`$t`)가 토큰 응답 전체를 쥐었다
     * (실측 2026-09-26). `expires_in` 이 소수면 만료 시각도 float 이라 같은 길로 샜다. `TokenSet::fromArray` 와
     * 같은 규칙으로 좁힌다 — access_token 은 강제변환하지 않는다(#481).
     */
    private function toTokenSet(#[\SensitiveParameter] AccessToken $t): TokenSet
    {
        $values = self::stringKeyed($t->getValues());
        $access = self::usableToken($t->getToken());
        if ($access === null) {
            throw new KeycloakAuthError('token response has no usable access_token');
        }
        $refresh = $t->getRefreshToken();
        $expires = self::toIntOrNull($t->getExpires());

        return new TokenSet(
            accessToken: $access,
            tokenType: isset($values['token_type']) ? self::toStr($values['token_type']) : 'Bearer',
            expiresIn: $expires !== null ? max(0, $expires - \time()) : 0,
            refreshToken: $refresh === null ? null : self::toStr($refresh),
            idToken: isset($values['id_token']) ? self::toStr($values['id_token']) : null,
            scope: isset($values['scope']) ? self::toStr($values['scope']) : null,
            expiresAt: $expires,
        );
    }

    /** 비어 있지 않은 문자열만 — `toStr` 처럼 숫자를 문자열로 바꾸지 않는다(쓸 수 없는 토큰이 된다). */
    private static function usableToken(#[\SensitiveParameter] mixed $v): ?string
    {
        return \is_string($v) && $v !== '' ? $v : null;
    }

    private static function toIntOrNull(mixed $v): ?int
    {
        return match (true) {
            \is_int($v) => $v,
            \is_float($v) => (int) $v,
            default => null,
        };
    }

    /** mixed 값을 문자열로 안전하게 좁힌다(신뢰된 OAuth 응답의 스칼라 값만 통과). */
    private static function toStr(mixed $v, string $default = ''): string
    {
        return match (true) {
            \is_string($v) => $v,
            \is_int($v), \is_float($v), \is_bool($v) => (string) $v,
            default => $default,
        };
    }

    /**
     * json_decode(..., true)/league getValues()의 배열은 키 타입이 array-key(int|string)로만 추론된다.
     * 신뢰된 OAuth/introspection 응답의 키는 항상 문자열이므로 정수 키(있다면)를 걸러 string-keyed로 좁힌다
     * (JwtValidator/JwksStore의 stringKeyed와 동일 패턴 — 로컬 헬퍼로 중복 유지, 공용화는 범위 밖).
     *
     * @param array<array-key, mixed> $a
     * @return array<string, mixed>
     */
    private static function stringKeyed(array $a): array
    {
        $out = [];
        foreach ($a as $k => $v) {
            if (is_string($k)) {
                $out[$k] = $v;
            }
        }

        return $out;
    }
}
